src/meta/src/manager/notification.rs

```diff
@@ -38,42 +38,113 @@ pub enum LocalNotification {
 /// Interval before retry when notify fail.
 const NOTIFY_RETRY_INTERVAL: Duration = Duration::from_micros(10);
 
+#[derive(Debug)]
+enum Target {
+    Frontend,
+    Compute,
+}
+
+#[derive(Debug)]
+struct Task {
+    target: Target,
+    callback_tx: Option<oneshot::Sender<NotificationVersion>>,
+    operation: Operation,
+    info: Info,
+}
+
 /// [`NotificationManager`] is used to send notification to frontends and compute nodes.
 pub struct NotificationManager {
-    core: Mutex<NotificationManagerCore>,
-    /// Sender used `Self::delete_sender` method.
+    core: Arc<Mutex<NotificationManagerCore>>,
+    /// Sender used to add a notification into the waiting queue.
+    task_tx: UnboundedSender<Task>,
+    /// Sender used in `Self::delete_sender` method.
     /// Tell `NotificationManagerCore` to skip some retry and delete senders.
-    tx: UnboundedSender<WorkerKey>,
+    stop_retry_tx: UnboundedSender<WorkerKey>,
 }
 
 pub type NotificationManagerRef = Arc<NotificationManager>;
 
 impl NotificationManager {
     pub fn new() -> Self {
-        let (tx, rx) = mpsc::unbounded_channel();
+        // notification waiting queue.
+        let (task_tx, mut task_rx) = mpsc::unbounded_channel::<Task>();
+        let (stop_retry_tx, stop_retry_rx) = mpsc::unbounded_channel();
+        let core = Arc::new(Mutex::new(NotificationManagerCore::new(stop_retry_rx)));
+        let core_clone = core.clone();
+
+        tokio::spawn(async move {
+            while let Some(task) = task_rx.recv().await {
+                let version = match task.target {
+                    Target::Frontend => {
+                        core.lock()
+                            .await
+                            .notify_frontend(task.operation, &task.info)
+                            .await
+                    }
+                    Target::Compute => {
+                        core.lock()
+                            .await
+                            .notify_compute(task.operation, &task.info)
+                            .await
+                    }
+                };
+                if let Some(tx) = task.callback_tx {
+                    tx.send(version).unwrap();
```

Comment:
Suggest warning here or it may panic on meta exiting.
```suggestion
                    if tx.send(version).is_err() { 
                        warn!(...);
                    }
```

Reply:
I think we can change this after ci error.